### DISCUSSION TITLE
rules compilation on windows, compiled rules used on iOS

### DISCUSSION THREAD

**Discussion by epouliquen** [NRules/NRules#264] [Q&A]
Hi,

  we are very interested in NRules possibilities, but our use case is very special and I would like your point of view on what is possible or not.

 We are developing a Unity3D app on iOS. This app is consuming data produced by  visual editors running on Windows desktop.
 We would like to use power of rules engine mechanism, and be able to design rules on windows, and use them in the iOS app.

 We have tried to use directly rules engines on the ios but of course we hit the 'compile' limitation for code generation on the type of aot platforms.

 My question is : could we eventually think about split NRules behavior in two steps : 
- on windows use it to compile rules and get a factory
- then find a way to serialize this factory instance (as dylib ?), put it on database
- then on iOS deserialize it and continue your "Getting Started" example , using compiled rules to avoid code generation on this platform

so, to illustrate it using your "Getting Started" example  :

 1. On Windows

```
//Load rules
var repository = new RuleRepository();
repository.Load(x => x.From(typeof(PreferredCustomerDiscountRule).Assembly));

//Compile rules
var factory = repository.Compile();
```

2. Find a way to save compiled rules code

3. and On iOS :

```
//retreive compiled rules code
var factory = new FactoryFromSomething()

//Create a working session
var session = factory.CreateSession();

//Load domain model
var customer = new Customer("John Doe") {IsPreferred = true};
var order1 = new Order(123456, customer, 2, 25.0);
var order2 = new Order(123457, customer, 1, 100.0);

//Insert facts into rules engine's memory
session.Insert(customer);
session.Insert(order1);
session.Insert(order2);

//Start match/resolve/act cycle
session.Fire();
```
Hope this strange use case is clear and hope you can help us :)

**Comment by snikolayev** [comment]
This is a very interesting use case. To be honest, I don't have much experience with AOT and runtimes that don't support dynamic compilation. Reading https://docs.microsoft.com/en-us/xamarin/ios/internals/limitations it seems like indeed, anything that directly or indirectly emits IL is not supported on iOS.
This also means there cannot be any intermediate format for representing executable rules other than actual fully compiled code. This is certainly possible, but I would imagine at the moment a LOT of work. Essentially somehow precompiling all Expressions and other parts of the engine that rely on dynamic IL gen.
One thing that seems clear is that there is a realization in the .net community that reflection/ILgen-based approach has its limitations, and where .net platform is going (AOT compilation, trimmed deployment, etc), there will only be more such use cases in the future. Given how many libraries and frameworks are relying on IL gen and reflection, hopefully the tools will get created that help with these problems. I'm pretty optimistic about the source generators that were recently added to .net. Still, precompiling lambda expressions for iOS is a gargantuan task, and currently there is no tooling that I know of that can do that (and I'm certainly not looking forward to hand-coding that).
So I would say - intriguing idea, potentially doable once/if the tooling emerges.

**Reply by epouliquen** [reply]
Thanks a lot for your answer.

I just would like to understand a first thing : where is the generated IL code ? is it contained in the factory after the "Compile" operation ? 

As e are working with Unity3d here, my first idea is to learn a bit more about their IL2CPP tool , which is used in the Mono to iOS build chain (c# -> (IL2CPP) -> c++ -> iOS binary code)

[https://katyscode.wordpress.com/2020/06/24/il2cpp-part-1/](https://katyscode.wordpress.com/2020/06/24/il2cpp-part-1/)

So as I wrote, but I'm probably wrong because I'm really a newbie in this domain, I would like to study the idea of convert your IL code to binaries through the IL2CPP chain, but puting IL rules into a .dylib that I can then call

**Reply by epouliquen** [reply]
Oh something strange, this thread [https://stackoverflow.com/questions/24977939/what-does-expression-compile-do-on-monotouch](https://stackoverflow.com/questions/24977939/what-does-expression-compile-do-on-monotouch)  (first answer) seems to say that Expression.Compile should work on AOT platforms, by switching from compilation to interpretation mode... As I see you don't use Emit in NRules code so it should work or I'm missing something. We are going to test deeper NRules compilation on iOS and I'll tell you compilation or untime errors if any, maybe we should find a solution.

**Reply by snikolayev** [reply]
NRules does not IL-gen directly, only indirectly via the use of Expressions, so your observations are correct - there is no direct use of Emit.
I'm also not experienced with the AOT platforms, so it may very well be that expressions fall back to interpretation mode on those platforms. Though, in that case, performance is likely going out of the window, so not sure if that's going to be ok. But the only way to find out is to try, so LMK what your testing reveals.
